### PR TITLE
Restrict root install jobs by default

### PR DESCRIPTION
**Pull request by hasheddan**
### Description of your changes

While the package manager currently restricts package controllers from running as default,
it does not place the same restrictions on package install jobs. This changes install jobs to be required to run as non-root by default, but can be disabled by passing the `--insecure-install-job` flag.

### How has this code been tested?

Integration tests. Manual testing to follow.

### Checklist

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [x] Updated any relevant [documentation] and [examples].
- [ ] Reported all new error conditions into the log or as an event, as
  appropriate.

For more about what we believe makes a pull request complete, see our
[definition of done].

[documentation]: https://github.com/crossplane/crossplane/tree/master/docs
[examples]: https://github.com/crossplane/crossplane/tree/master/cluster/examples
[definition of done]: https://github.com/crossplane/crossplane/tree/master/design/one-pager-definition-of-done.md

### CODE REVIEW

**Comment by hasheddan** [comment]
@lukeweber when unpacking happens the image that is used is the Crossplane image itself, because everything is bundled into a single binary. The user doesn't actually get a say in the `Job` that is created, so if you trust Crossplane then the current status would be okay. However, I don't see a reason for why the Crossplane pod (either for `core`, `package-manager`, or `unpack`) should ever need to run as root, so I think it would be best practice to make it not. Then the security context changes on the `Job` are essentially just a check to make sure that the install is never running as root.

**Comment by negz** [comment]
@hasheddan @lukeweber is this something that needs to land for Crossplane 0.11? If so, @jbw976 do you feel familiar enough with this part of the codebase to review?

File: cluster/charts/crossplane/README.md [hunk 1]
```diff
@@ -314,7 +314,8 @@ and their default values.
...
 | `forceImagePullPolicy`           | Force the named ImagePullPolicy on Package install and containers | `""`
 | `insecureAllowAllApigroups`      | Enable core Kubernetes API group permissions for Packages. When enabled, Packages may declare dependency on core Kubernetes API types.) | `false` |
-| `insecurePassFullDeployment`     | Enable packages to pass their full deployment, including security context. When omitted, Packages deployments will have security context removed and all containers will have `allowPrivilegeEscalation` set to false. | `false` |
+| `insecurePassFullDeployment`     | Enable packages to pass their full deployment, including security context. When omitted, Package controller deployments will have security context removed and all containers will have `privileged` and `allowPrivilegeEscalation` set to `false`, and `runAsNonRoot` set to `true`. | `false` |
+| `insecureInstallJob`     | Enable package install jobs to run as root. When omitted, Package install jobs will have security context removed and all containers containers will have `privileged` and `allowPrivilegeEscalation` set to `false`, and `runAsNonRoot` set to `true`. | `false` |
```
Comment (jbw976) [inline cluster/charts/crossplane/README.md:318]: typo: duplicated `containers containers`?

File: cmd/crossplane/package/manage/manage.go [hunk 2]
```diff
@@ -50,7 +51,8 @@ func FromKingpin(cmd *kingpin.CmdClause) *Command {
...
-	cmd.Flag("insecure-pass-full-deployment", "Enable packagess to pass their full deployment, including security context. When omitted, Packages deployments will have security context removed and all containers will have allowPrivilegeEscalation set to false.").Default("false").BoolVar(&c.PassFullDeployment)
+	cmd.Flag("insecure-pass-full-deployment", "Enable packages to pass their full deployment, including security context. When omitted, Package controller deployments will have security context removed and all containers will have privileged and allowPrivilegeEscalation set to false, and runAsNonRoot set to true.").Default("false").BoolVar(&c.PassFullDeployment)
+	cmd.Flag("insecure-install-job", "Enable package install jobs to run as root. When omitted, Package install jobs will have security context removed and all containers containers will have privileged and allowPrivilegeEscalation set to false, and runAsNonRoot set to true.").Default("false").BoolVar(&c.InsecureInstallJob)
```
Comment (jbw976) [inline cmd/crossplane/package/manage/manage.go:55]: typo: `containers containers` shows up in this text too

File: docs/getting-started/install.md [hunk 3]
```diff
@@ -321,7 +321,8 @@ and their default values.
...
 | `forceImagePullPolicy`           | Force the named ImagePullPolicy on Package install and containers | `""`
 | `insecureAllowAllApigroups`      | Enable core Kubernetes API group permissions for Packages. When enabled, Packages may declare dependency on core Kubernetes API types.) | `false` |
-| `insecurePassFullDeployment`     | Enable packages to pass their full deployment, including security context. When omitted, Packages deployments will have security context removed and all containers will have `allowPrivilegeEscalation` set to false. | `false` |
+| `insecurePassFullDeployment`     | Enable packages to pass their full deployment, including security context. When omitted, Package controller deployments will have security context removed and all containers will have `privileged` and `allowPrivilegeEscalation` set to `false`, and `runAsNonRoot` set to `true`. | `false` |
+| `insecureInstallJob`     | Enable package install jobs to run as root. When omitted, Package install jobs will have security context removed and all containers containers will have `privileged` and `allowPrivilegeEscalation` set to `false`, and `runAsNonRoot` set to `true`. | `false` |
```
Comment (jbw976) [inline docs/getting-started/install.md:325]: typo: `containers containers`

**Comment by hasheddan** [comment]
@jbw976 good catch! I believe those are fixed now :+1:

File: cluster/charts/crossplane/README.md [hunk 4]
```diff
@@ -314,7 +314,8 @@ and their default values.
...
 | `forceImagePullPolicy`           | Force the named ImagePullPolicy on Package install and containers | `""`
 | `insecureAllowAllApigroups`      | Enable core Kubernetes API group permissions for Packages. When enabled, Packages may declare dependency on core Kubernetes API types.) | `false` |
-| `insecurePassFullDeployment`     | Enable packages to pass their full deployment, including security context. When omitted, Packages deployments will have security context removed and all containers will have `allowPrivilegeEscalation` set to false. | `false` |
+| `insecurePassFullDeployment`     | Enable packages to pass their full deployment, including security context. When omitted, Package controller deployments will have security context removed and all containers will have `privileged` and `allowPrivilegeEscalation` set to `false`, and `runAsNonRoot` set to `true`. | `false` |
+| `insecureInstallJob`     | Enable package install jobs to run as root. When omitted, Package install jobs will have security context removed and all will have `privileged` and `allowPrivilegeEscalation` set to `false`, and `runAsNonRoot` set to `true`. | `false` |
```
Comment (jbw976) [inline cluster/charts/crossplane/README.md:318]: is it missing `containers` completely now? `all will` -> `all containers will`, like it says in cmd/crossplane/package/manage/manage.go`

File: docs/getting-started/install.md [hunk 5]
```diff
@@ -321,7 +321,8 @@ and their default values.
...
 | `forceImagePullPolicy`           | Force the named ImagePullPolicy on Package install and containers | `""`
 | `insecureAllowAllApigroups`      | Enable core Kubernetes API group permissions for Packages. When enabled, Packages may declare dependency on core Kubernetes API types.) | `false` |
-| `insecurePassFullDeployment`     | Enable packages to pass their full deployment, including security context. When omitted, Packages deployments will have security context removed and all containers will have `allowPrivilegeEscalation` set to false. | `false` |
+| `insecurePassFullDeployment`     | Enable packages to pass their full deployment, including security context. When omitted, Package controller deployments will have security context removed and all containers will have `privileged` and `allowPrivilegeEscalation` set to `false`, and `runAsNonRoot` set to `true`. | `false` |
+| `insecureInstallJob`     | Enable package install jobs to run as root. When omitted, Package install jobs will have security context removed and all will have `privileged` and `allowPrivilegeEscalation` set to `false`, and `runAsNonRoot` set to `true`. | `false` |
```
Comment (jbw976) [inline docs/getting-started/install.md:325]: same, `all will` -> `all containers will`

File: cluster/charts/crossplane/README.md (hunk 4 above)
Comment (hasheddan) [reply cluster/charts/crossplane/README.md:318]: hmm did find and replace here so not sure how it worked for the `cmd` but not `docs`. Oh well, will update again :man_facepalming:

**Pull request merged**